crypto/x509/v3_purp.c

```diff
@@ -305,44 +305,62 @@ int X509_supported_extension(X509_EXTENSION *ex)
     return 0;
 }
 
-static int setup_dp(X509 *x, DIST_POINT *dp)
+/* return 1 on success, 0 if x is invalid, -1 on (internal) error */
+static int setup_dp(const X509 *x, DIST_POINT *dp)
 {
     const X509_NAME *iname = NULL;
     int i;
 
-    if (dp->reasons) {
+    if (dp->distpoint == NULL && sk_GENERAL_NAME_num(dp->CRLissuer) <= 0) {
```

Comment:
I'm currently not sure if == 0 is a problem or not. Below you seem to say that <= 1 is acceptable, not == 1. You can change it to `dp->CRLissuer == NULL`

Reply:
I currently can't see how == 0 can make any sense, so I guess it's fine.

Reply:
> I'm currently not sure if == 0 is a problem or not.

It is - according to the 2nd paragraph of section 4.2.1.13., if distributionPoint is missing the cRLissuer must contain at least one entry:

> If the certificate issuer is not the CRL
>    issuer, then the cRLIssuer field MUST be present and contain the Name
>    of the CRL issuer.  If the certificate issuer is also the CRL issuer,
>    then conforming CAs MUST omit the cRLIssuer field and MUST include
>    the distributionPoint field


> Below you seem to say that <= 1 is acceptable, not == 1

That refers to a different requirement further down in the same section:

> The
>    DistributionPointName MUST NOT use the nameRelativeToCRLIssuer
>    alternative when cRLIssuer contains more than one distinguished name.


> You can change it to `dp->CRLissuer == NULL`

Not really, because as just mentioned if dp->distpoint == NULL then at least one cRLissuer element must be present.

Reply:
So I've added some new checks to x509 that should check the same:
https://github.com/kroeckx/x509lint/blob/2e2613e5953d37fb6ffb27cff94cb55167c6c533/checks.c#L1164-L1177
https://github.com/kroeckx/x509lint/blob/2e2613e5953d37fb6ffb27cff94cb55167c6c533/checks.c#L1206-L1209

Reply:
> So I've added some new checks to x509 that should check the same:

Nice.

> https://github.com/kroeckx/x509lint/blob/2e2613e5953d37fb6ffb27cff94cb55167c6c533/checks.c#L1164-L1177

The 
```
			if (dp->distpoint == NULL && dp->CRLissuer == NULL)
			{
				SetError(ERR_INVALID_CRL_DIST_POINT);
			}
			if (dp->CRLissuer != NULL)
			{
				if (sk_GENERAL_NAME_num(dp->CRLissuer) == 0)
				{
					SetError(ERR_INVALID_CRL_DIST_POINT);
				}
```
does not cover the case that CRLissuer is present but empty. So should be as I use in this PR:
```
if (dp->distpoint == NULL &&  sk_GENERAL_NAME_num(dp->CRLissuer) <= 0)
			{
				SetError(ERR_INVALID_CRL_DIST_POINT);
			}
```

> https://github.com/kroeckx/x509lint/blob/2e2613e5953d37fb6ffb27cff94cb55167c6c533/checks.c#L1206-L1209

The `if (dp->CRLissuer != NULL && sk_GENERAL_NAME_num(dp->CRLissuer) != 1)`
can be simplified to `if (sk_GENERAL_NAME_num(dp->CRLissuer) <= 1)`.

Reply:
> does not cover the case that CRLissuer is present but empty.

I'm not sure what you're saying, since that seems to be exactly what's checked.

Reply:
According to RFC 5280 and
```
        ASN1_IMP_SEQUENCE_OF_OPT(DIST_POINT, CRLissuer, GENERAL_NAME, 2)
```
the CRLissuer list is optional, so can be absent
(`dp->CRLissuer == NULL` or equivalently `sk_GENERAL_NAME_num(dp->CRLissuer) == -1`), 
and at least according to the above OpenSSL declaration it can be present but empty 
(`sk_GENERAL_NAME_num(dp->CRLissuer) == 0`).
Of course, but cases mean that no CRLissuer is included.

I caught both cases by `sk_GENERAL_NAME_num(dp->CRLissuer) <= 0`.

Reply:
Ah yes, together with the guard `if (dp->CRLissuer != NULL)`
your code does cover the case that the CRLissuer field is present but empty.